### PR TITLE
feat: add OOG error handling for `CALLCODE`, `DELEGATECALL` and `STATICCALL` to `ErrorOOGCall`

### PR DESCRIPTION
**Pull request by silathdiir**
Close https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1139

Spec PR https://github.com/privacy-scaling-explorations/zkevm-specs/pull/368 has been merged.

### Summary

1. Merge enum value `OogError::CallCode`, `DelegateCall` and `StaticCall` into `OogError::Call`.

2. Merge enum value `ExecutionState::ErrorOutOfGasCALLCODE`, `ErrorOutOfGasDELEGATECALL` and `ErrorOutOfGasSTATICCALL` into `ExecutionState::ErrorOutOfGasCALL`.

3. Update bus-mapping and zkevm-circuit `error_oog_call` to handle CALLCODE, DELEGATECALL and STATICCALL opcodes.

4. Update both root and internal test cases in circuit.

### CODE REVIEW

**Review by lispc — APPROVED** [review]

File: zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs [hunk 1]
```diff
@@ -34,23 +43,37 @@ pub(crate) struct ErrorOOGCallGadget<F> {
...
         let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
         let is_static = cb.call_context(None, CallContextFieldTag::IsStatic);
-        let call_gadget = CommonCallGadget::construct(cb, 1.expr(), 0.expr(), 0.expr());
+
+        let call_gadget = CommonCallGadget::construct(
+            cb,
+            is_call.expr(),
+            is_callcode.expr(),
+            is_delegatecall.expr(),
+        );
```
Comment (han0110) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs:76]: Just realized that we might need to also pass `is_static` to `CommonCallGadget` here, to make sure the `has_value` is `0` when it's `STATICCALL`, the changes in `CommonCallGadget` will be:

https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/e19ba01ecf343e7a331eae8d268070d84e89aeea/zkevm-circuits/src/evm_circuit/util/common_gadget.rs#L517-L521

```diff
     let has_value = select::expr( 
-        is_delegatecall.expr(), 
+        is_delegatecall.expr() + is_staticcall.expr(), 
         0.expr(), 
         1.expr() - value_is_zero.expr(), 
     ); 
```


Also in line 87 the second argument of `gas_cost_expr` should be passed with `is_call`.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs [hunk 2]
```diff
@@ -34,23 +43,37 @@ pub(crate) struct ErrorOOGCallGadget<F> {
...
-            opcode.expr(),
-            OpcodeId::CALL.expr(),
+            "ErrorOutOfGasCall opcode should be CALL, CALLCODE, DELEGATECALL or STATICCALL",
+            is_call.expr() + is_callcode.expr() + is_delegatecall.expr() + is_staticcall.expr(),
+            1.expr(),
         );
 
         let rw_counter_end_of_reversion = cb.query_cell();
         let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
         let is_static = cb.call_context(None, CallContextFieldTag::IsStatic);
```
Comment (han0110) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs:69]: This seems not used, perhaps we could remove it. (gas calculation doesn't depends on if it's read-only context).

**Review by han0110 — COMMENTED** [review]
Overall LGTM.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs (hunk 1 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs]: Make below fixes in 37e1e1e:

1. Add `is_staticcall` parameter to `CommonCallGadget::construct`.

2. Move `require_equal(is_call + is_callcode + is_delegatecall + is_staticcall, 1)` constraint from both `CallOpGadget` and `ErrorOOGCallGadget` into CommonCallGadget.

3. Fix `has_value = select(is_delegatecall,` to `has_value = select(is_delegatecall + is_staticcall,` in CommonCallGadget.

4. Fix to pass `is_call` (to replace `1.expr()`) into `call_gadget.gas_cost_expr` in ErrorOGGCallGadget.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs (hunk 2 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs:60]: I try to reference the code of `CallOpGadget` and find a constraint of `has_value` and `is_static` as below. I copied it into `ErrorOOGCallGadget` in 8622af9.
```
cb.condition(call_gadget.has_value.expr(), |cb| {
    cb.require_zero(
        "CALL with value must not be in static call stack",
        is_static.expr(),
    );
});
```

File: zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs (hunk 1 above)
Comment (han0110) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs]: Looks much better now!

**Review by han0110 — APPROVED** [review]

**Pull request merged**